pkg/generate/appjson/appjson.go

```diff
@@ -0,0 +1,388 @@
+
+	deployapi "github.com/openshift/origin/pkg/deploy/api"
+	"github.com/openshift/origin/pkg/generate/app"
+	templateapi "github.com/openshift/origin/pkg/template/api"
+	"github.com/openshift/origin/pkg/util/docker/dockerfile"
+)
+
+type EnvVarOrString struct {
+	Value  string
+	EnvVar *EnvVar
+}
+
+type EnvVar struct {
+	Description string
+	Generator   string
+	Value       string
+	Required    bool
+	Default     interface{}
+}
+
+func (e *EnvVarOrString) UnmarshalJSON(data []byte) error {
+	if len(data) < 2 {
+		return nil
+	}
+	if data[0] == '"' {
+		e.Value = string(data[1 : len(data)-1])
+		return nil
+	}
+	e.EnvVar = &EnvVar{}
+	return json.Unmarshal(data, e.EnvVar)
+}
+
+type Formation struct {
+	Quantity int32
+	Size     string
+	Command  string
+}
+
+type Buildpack struct {
+	URL string `json:"url"`
+}
+
+type AppJSON struct {
+	Name        string
+	Description string
+	Keywords    []string
+	Repository  string
+	Website     string
+	Logo        string
+	SuccessURL  string `json:"success_url"`
+	Scripts     map[string]string
+	Env         map[string]EnvVarOrString
+	Formation   map[string]Formation
+	Image       string
+	Addons      []string
+	Buildpacks  []Buildpack
+}
+
+type Generator struct {
+	LocalPath string
+	Name      string
+}
+
+// Generate accepts a path to an app.json file and generates a template from it
+func (g *Generator) Generate(body []byte) (*templateapi.Template, error) {
+	appJSON := &AppJSON{}
+	if err := json.Unmarshal(body, appJSON); err != nil {
+		return nil, err
+	}
+
+	glog.V(4).Infof("app.json: %#v", appJSON)
+
+	warnings := make(map[string][]string)
+	warnUnusableAppJSONElements("app.json", appJSON, warnings)
+
+	name := g.Name
+	if len(name) == 0 && len(g.LocalPath) > 0 {
+		name = filepath.Base(g.LocalPath)
+	}
+
+	template := &templateapi.Template{}
+	template.Name = name
+	template.Annotations = make(map[string]string)
+	template.Annotations["openshift.io/website"] = appJSON.Website
+	template.Annotations["k8s.io/display-name"] = appJSON.Name
+	template.Annotations["k8s.io/description"] = appJSON.Description
+	template.Annotations["tags"] = strings.Join(appJSON.Keywords, ",")
+	template.Annotations["iconURL"] = appJSON.Logo
+
+	// create parameters and environment for containers
+	allEnv := make(app.Environment)
+	for k, v := range appJSON.Env {
+		if v.EnvVar != nil {
+			allEnv[k] = fmt.Sprintf("${%s}", k)
+		}
+	}
+	envVars := allEnv.List()
+	for _, v := range envVars {
+		env := appJSON.Env[v.Name]
+		if env.EnvVar == nil {
+			continue
+		}
+		e := env.EnvVar
+		displayName := v.Name
+		displayName = strings.Join(strings.Split(strings.ToLower(displayName), "_"), " ")
+		displayName = strings.ToUpper(displayName[:1]) + displayName[1:]
+		param := templateapi.Parameter{
+			Name:        v.Name,
+			DisplayName: displayName,
+			Description: e.Description,
+			Value:       e.Value,
+		}
+		switch e.Generator {
+		case "secret":
+			param.Generate = "expression"
+			param.From = "[a-zA-Z0-9]{14}"
+		}
```

Comment:
I want app.json to work - it won't today without support for generators.

On Tue, May 10, 2016 at 3:20 PM, Ben Parees <notifications@github.com>
wrote:

> In pkg/generate/appjson/appjson.go
> <https://github.com/openshift/origin/pull/8819#discussion_r62735114>:
>
> > +		}
> > +		e := env.EnvVar
> > +		displayName := v.Name
> > +		displayName = strings.Join(strings.Split(strings.ToLower(displayName), "_"), " ")
> > +		displayName = strings.ToUpper(displayName[:1]) + displayName[1:]
> > +		param := templateapi.Parameter{
> > +			Name:        v.Name,
> > +			DisplayName: displayName,
> > +			Description: e.Description,
> > +			Value:       e.Value,
> > +		}
> > +		switch e.Generator {
> > +		case "secret":
> > +			param.Generate = "expression"
> > +			param.From = "[a-zA-Z0-9]{14}"
> > +		}
>
> given that generators are going away in k8s templates, do we really want
> to invest more in them here?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8819/files/89ca3f79b38ea3ffc3fd3165cec0b0394ad8bbad#r62735114>
>